Dl/tracing intro (#150)
mlflow/mlflow-website: website/blog/2025-02-26-tracing-introduction/index.mdx

If you use tracing across multiple different projects and tasks, you might want to organize the traces into separate groups.

The easiest way to organize your traces is to separate them into [*experiments*](https://mlflow.org/docs/latest/tracking.html#experiments). Each experiment has its own traces tab, which displays the traces for that experiment.

You can create an experiment in the UI (With the "+" button next to "Experiments"), with the MLflow CLI, or with Python. Let's create a new experiment called "Quickstart" and log a trace.

```python
mlflow.set_experiment("quickstart")

completion = client.chat.completions.create(
    model="gpt-4o-mini",
    messages=[
        {"role": "system", "content": "You are a helpful assistant."},
        {
            "role": "user",
            "content": "What is an MLflow tracking server?"
        }
    ]
)
```

We can now find this trace in the "Traces" tab in the "quickstart" experiment.

![Trace in Experiment](/img/blog/tracing-intro/03_experiment.png)

The `set_experiment` function specifies which experiment traces should be logged to, creating it if it does not exist, so the code snippet above created a new "quickstart" experiment.

> You can also organize your traces with [tags](https://mlflow.org/docs/latest/llms/tracing/index.html#trace-tags) and [runs](https://mlflow.org/docs/latest/llms/tracing/index.html#q-how-can-i-associate-a-trace-with-an-mlflow-run).

Reviewer: ```suggestion
You can also organize your traces with [tags](https://mlflow.org/docs/latest/tracing#trace-tags) and [runs](https://mlflow.org/docs/latest/tracing#q-how-can-i-associate-a-trace-with-an-mlflow-run).
```
(updates for the new urls - the old ones have redirects, but we should use the new ones)

## Tracing Other Providers

> Our quickstart example focused on OpenAI, but MLflow supports automatic tracing of [many different AI providers and frameworks](https://mlflow.org/docs/latest/llms/tracing/index.html#automatic-tracing). The approach is the same: just add the line `mlflow.<provider>.autolog` to your notebook or script.

Reviewer: ```suggestion
Our quickstart example focused on OpenAI, but MLflow supports automatic tracing of [many different AI providers and frameworks](https://mlflow.org/docs/latest/tracing#automatic-tracing). The approach is the same: just add the line `mlflow.<provider>.autolog` to your notebook or script.
```

> Here are a few examples. See [here](https://mlflow.org/docs/latest/llms/tracing/index.html#automatic-tracing) for the full list of supported providers.

Reviewer: ```suggestion
Here are a few examples. See [here](https://mlflow.org/docs/latest/tracing#automatic-tracing) for the full list of supported providers.
```

prompt_template = PromptTemplate.from_template(
    "Explain the following MLflow concept at the specified technical level. "
    "For 'beginner', use simple analogies and avoid complex terms. "
    "For 'intermediate', include more technical details and some code examples. "
    "For 'advanced', go deep into implementation details and provide comprehensive explanations. "
    "Technical level: {level}. Question: {question}"
)

chain = prompt_template | llm | StrOutputParser()

chain.invoke(
    {
        "level": "beginner",
        "question": "How do MLflow tracking servers help with experiment management?",
    }
)
```

This example LangChain chain includes multiple components:

- The `PromptTemplate`, which assembles the prompt based on the user input
- The `ChatOpenAI` model, which is used to call the OpenAI `gpt-4o-mini` model
- The `StrOutputParser`, which returns the final answer to the user's query as a string

We can see each of these components in the MLflow UI, nested under the parent `RunnableSequence` chain.

![LangChain Tracing](/img/blog/tracing-intro/05_langchain.png)

  </TabItem>
  <TabItem value="ollama" label="Ollama">

> [Ollama](ollama.ai) is a tool for running open source AI models locally. You can enable automatic tracing of Ollama models via Ollama's [OpenAI-compatible API](https://github.com/ollama/ollama/blob/main/docs/openai.md) and MLflow's OpenAI autologging. You just need to set the base URL to your Ollama REST endpoint.

Reviewer: ```suggestion
[Ollama](https://ollama.com/) is a tool for running open source AI models locally. You can enable automatic tracing of Ollama models via Ollama's [OpenAI-compatible API](https://github.com/ollama/ollama/blob/main/docs/openai.md) and MLflow's OpenAI autologging. You just need to set the base URL to your Ollama REST endpoint.
```

  
This pattern should work with any provider that offers an OpenAI-compatible endpoint, even those that are not explicitly referenced in the docs.

Here's how it works for Ollama:

1. First, run the Ollama server with your desired model.

```bash
ollama run phi3:latest
```

2. Configure the OpenAI client, setting the `base_url` to the Ollama OpenAI-compatible endpoint.

```python
from openai import OpenAI

client = OpenAI(
    base_url="http://localhost:11434/v1",  # The local Ollama REST endpoint
    api_key="dummy",  # Required to instantiate OpenAI client, it can be a random string
)
```

3. Enable MLflow OpenAI autologging and query the model

```python
mlflow.openai.autolog()

completion = client.chat.completions.create(
    model="phi3:latest",
    messages=[
        {"role": "system", "content": "You are a helpful assistant."},
        {"role": "user", "content": "What is an MLflow tracking server?"}
    ]
)
```

Here is the trace of the Ollama model call in the MLflow UI.

![Ollama Tracing](/img/blog/tracing-intro/06_ollama.png)
  
  </TabItem>
</Tabs>

## Conclusion: Effective LLM tracing with one line of code

> In this guide, you have learned how to use MLflow's autologging capabilities to get a complete AI observability solution with a single line of code. If you are using one of the [many GenAI frameworks/providers](https://mlflow.org/docs/latest/llms/tracing/index.html#automatic-tracing) for which MLflow offers automatic tracing—including any providers with OpenAI-compatible endpoints—automatic logging is the easiest way to visualize and debug your AI application behavior. All you need is `mlflow.<provider>.autolog()`.

Reviewer: ```suggestion
In this guide, you have learned how to use MLflow's autologging capabilities to get a complete AI observability solution with a single line of code. If you are using one of the [many GenAI frameworks/providers](https://mlflow.org/docs/latest/tracing#automatic-tracing) for which MLflow offers automatic tracing—including any providers with OpenAI-compatible endpoints—automatic logging is the easiest way to visualize and debug your AI application behavior. All you need is `mlflow.<provider>.autolog()`.
```

### Next Steps

Autologging is a great place to start with MLflow tracing, but you may need more flexibility in how you collect and use traces as you develop more complex GenAI applications. Furthermore, MLflow includes many tools for working with GenAI applications beyond tracing.

> - For a longer conceptual introduction to tracing, read [this guide](https://mlflow.org/docs/latest/llms/tracing/overview.html) on tracing concepts.

Reviewer: ```suggestion
- For a longer conceptual introduction to tracing, read [this guide](https://mlflow.org/docs/latest/tracing/) on tracing concepts.
```

> - MLflow traces can provide an excellent source of data for evaluation, SME review, fine-tuning, and more. Learn about searching and retrieving trace data [here](https://mlflow.org/docs/latest/llms/tracing/search-traces.html).

Reviewer: ```suggestion
- MLflow traces can provide an excellent source of data for evaluation, SME review, fine-tuning, and more. Learn about searching and retrieving trace data [here](https://mlflow.org/docs/latest/tracing/api/search).
```

- MLflow provides [LLM evaluation functionality](https://mlflow.org/docs/latest/llms/llm-evaluate/index.html) for running structured experiments with your AI models and applications.

> - You can add tracing to your own AI applications with the tracing [fluent APIs](https://mlflow.org/docs/latest/llms/tracing/index.html#tracing-fluent-apis) and [client APIs](https://mlflow.org/docs/latest/llms/tracing/index.html#tracing-client-apis). You can also [add tracing to libraries and frameworks](https://mlflow.org/blog/custom-tracing) that do not (yet) have autologging support.

Reviewer: ```suggestion
- You can add tracing to your own AI applications with the tracing [fluent APIs](https://mlflow.org/docs/latest/tracing/api/manual-instrumentation) and [client APIs](https://mlflow.org/docs/latest/tracing/api/client). You can also [add tracing to libraries and frameworks](https://mlflow.org/blog/custom-tracing) that do not (yet) have autologging support.
```
